pkg/controller/snapshot_controller.go

```diff
@@ -0,0 +1,773 @@
+		if err != nil {
+			glog.V(4).Infof("updating VolumeSnapshotContent[%s] error status failed %v", newContent.Name, err)
+			return err
+		}
+		_, err = ctrl.storeContentUpdate(newContent)
+		if err != nil {
+			glog.V(4).Infof("updating VolumeSnapshotContent[%s] error status: cannot update internal cache %v", newContent.Name, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (ctrl *csiSnapshotController) checkandUpdateSnapshotStatusOperation(snapshot *crdv1.VolumeSnapshot, content *crdv1.VolumeSnapshotContent) (*crdv1.VolumeSnapshot, error) {
+	status, _, err := ctrl.handler.GetSnapshotStatus(content)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check snapshot status %s with error %v", snapshot.Name, err)
+	}
+
+	newSnapshot, err := ctrl.updateSnapshotStatus(snapshot, status, time.Now(), IsSnapshotBound(snapshot, content))
+	if err != nil {
+		return nil, err
+	}
+	return newSnapshot, nil
+}
+
+// The function goes through the whole snapshot creation process.
+// 1. Trigger the snapshot through csi storage provider.
+// 2. Update VolumeSnapshot status with creationtimestamp information
+// 3. Create the VolumeSnapshotContent object with the snapshot id information.
+// 4. Bind the VolumeSnapshot and VolumeSnapshotContent object
+func (ctrl *csiSnapshotController) createSnapshotOperation(snapshot *crdv1.VolumeSnapshot) (*crdv1.VolumeSnapshot, error) {
+	glog.Infof("createSnapshot: Creating snapshot %s through the plugin ...", snapshotKey(snapshot))
+
+	class, err := ctrl.GetClassFromVolumeSnapshot(snapshot)
+	if err != nil {
+		glog.Errorf("createSnapshotOperation failed to getClassFromVolumeSnapshot %s", err)
+		return nil, err
+	}
+	volume, err := ctrl.getVolumeFromVolumeSnapshot(snapshot)
+	if err != nil {
+		glog.Errorf("createSnapshotOperation failed to get PersistentVolume object [%s]: Error: [%#v]", snapshot.Name, err)
+		return nil, err
+	}
+
+	// Create VolumeSnapshotContent name
+	contentName := GetSnapshotContentNameForSnapshot(snapshot)
+
+	// Resolve snapshotting secret credentials.
+	snapshotterSecretRef, err := GetSecretReference(snapshotterSecretNameKey, snapshotterSecretNamespaceKey, class.Parameters, contentName, snapshot)
+	if err != nil {
+		return nil, err
+	}
+	snapshotterCredentials, err := GetCredentials(ctrl.client, snapshotterSecretRef)
+	if err != nil {
+		return nil, err
+	}
+
+	driverName, snapshotID, timestamp, csiSnapshotStatus, err := ctrl.handler.CreateSnapshot(snapshot, volume, class.Parameters, snapshotterCredentials)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to take snapshot of the volume, %s: %q", volume.Name, err)
+	}
+	glog.Infof("Create snapshot driver %s, snapshotId %s, timestamp %d, csiSnapshotStatus %v", driverName, snapshotID, timestamp, csiSnapshotStatus)
+
+	// Update snapshot status with timestamp
+	newSnapshot, err := ctrl.updateSnapshotStatus(snapshot, csiSnapshotStatus, time.Unix(0, timestamp), false)
```

Comment:
If update fails, you end up with a snapshot in the storage backend and without `VolumeSnapshotContent` for it.

Reply:
that's a good point. We can add try to update snapshot a few times similar to trying to create VolumeSnapshotContent a few times.

I have been thinking how to handle failures and this (not) retry policy. Maybe it is better to differentiate some failures.
1. If CSI driver fail to create the snapshot and returns the error, the controller will set the error state and not try to issue the creation again.
2. If CSI driver creates the snapshot successfully, but controller fails to update the status or create snapshotContent, the controller can resync the API and try to update the status again. 

I will see how controller can handle these.